### ISSUE TITLE
[FR] `MagicBox` operator

### ISSUE THREAD

**Issue by ordabayevy** [pyro-ppl/funsor#542]
This proposes to add `MagicBox` funsor that eagerly evaluates to `exp(x - x.detach())`. `MagicBox` operator would be useful for implementing https://github.com/pyro-ppl/pyro/issues/2869.

**Comment by eb8680** [comment]
Note that we already add Dice factors to non-reparametrized samples in [`Distribution.unscaled_sample`](https://github.com/pyro-ppl/funsor/blob/master/funsor/distribution.py#L237) and [`Tensor.unscaled_sample`](https://github.com/pyro-ppl/funsor/blob/master/funsor/tensor.py#L406). IIUC the proposal in https://github.com/pyro-ppl/pyro/issues/2869 would correspond to different implementations of `Funsor.sample` for different gradient estimators?

**Comment by ordabayevy** [comment]
> Note that we already add Dice factors to non-reparametrized samples in [`Distribution.unscaled_sample`](https://github.com/pyro-ppl/funsor/blob/master/funsor/distribution.py#L237)

Thanks for pointing that out.

As I mentioned in https://github.com/pyro-ppl/pyro/pull/2893 I can't find where [`Distribution.unscaled_sample`](https://github.com/pyro-ppl/funsor/blob/master/funsor/distribution.py#L237) is called in `contrib.funsor.infer.Trace_ELBO`.
